### PR TITLE
feat(binder): add const case-when evaluation optimization during binding

### PR DESCRIPTION
**Pull request by xzhseh**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

As titled, the optimization goes to case-when expression in below format, e.g.,

```sql
select case <constant> 
    when <constant> then <expr_1>
    when <constant> then <expr_2>
    ... (potentially lots of other arms with constant condition clause)
    else <expr_fallback> end;
```

related: https://github.com/risingwavelabs/risingwave/pull/14586#issue-2082757842.

And since I'm new to optimizer, feel free to point out any part I may have missed and any suggestion is welcomed. 😄

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/frontend/src/binder/expr/mod.rs [hunk 1]
```diff
@@ -469,6 +469,60 @@ impl Binder {
         Ok(func_call.into())
     }
 
+    /// The optimization check for the following case-when expression pattern
+    /// e.g., select case 1 when (...) then (...) else (...) end;
+    fn check_constant_case_when_optimization(
```
Comment (xzhseh) [inline src/frontend/src/binder/expr/mod.rs:474]: The reason I create another check function is that we can't check all the optimizable case-when patterns within only one pass, and they indeed have subtle but significant difference, which may not be generalized.

File: src/frontend/src/binder/expr/mod.rs [hunk 2]
```diff
@@ -536,6 +596,26 @@ impl Binder {
...
+
+        if constant_case_when_flag {
+            // Here we reuse the `ConstantLookup` as the `FunctionCall`
+            // to avoid creating a dummy `ConstCaseWhenEval` expression type
+            // since we do not need to go through backend
+            return Ok(FunctionCall::new(
+                ExprType::ConstantLookup,
+                constant_case_when_eval_inputs,
+            )?
+            .into());
```
Comment (chenzl25) [inline src/frontend/src/binder/expr/mod.rs:617]: Can we just return `constant_case_when_eval_inputs[0]` here? Then we don't even need to use a rule to convert it later.

Comment (xzhseh) [reply src/frontend/src/binder/expr/mod.rs:617]: Then what type should we return, I doubt FunctionCall would work 🤔, also Literal may not be the appropriate one.

Comment (chenzl25) [reply src/frontend/src/binder/expr/mod.rs:617]: This method returns an `ExprImpl`, an enum that could be any expression type.

Comment (xzhseh) [reply src/frontend/src/binder/expr/mod.rs:617]: Got it!
Then I think return a bound expression should work, and I'll remove the relevant rules. 😄

**Review by chenzl25 — APPROVED** [review]
LGTM

**Pull request merged**